## Prise en main

In [1]:
x = 2434+33476
y = 1+3 

In [ ]:
x

In [2]:
print(x)

35910


In [ ]:
x
print(x)

In [ ]:
y

In [ ]:
y
print(x)

In [ ]:
print(y)
x

In [ ]:
print?

In [ ]:
print

In [34]:
import numpy as np
import scipy
import matplotlib as plt
import sklearn

# Exercices

## Exercice 1 : norme d'une liste
Ecrire une fonction Python qui calcule la norme Euclidienne d'un vecteur de dimension arbitraire définie sous forme de liste.

In [5]:
vecteur = [5,3,np.sqrt(2)]

In [9]:
def normeEuclidienne(list):
    ret = 0
    for i in range(0,len(list)):
        ret = ret + list[i] ** 2
    ret = np.sqrt(ret)
    return ret

In [10]:
print(normeEuclidienne(vecteur))

6.0


Citez des avantages des listes en Python ? [A compléter]

1. 
2. 
3. 
4. 

## Exercice 1bis : matrice Numpy

* Ecrire une méthode créant une matrice, sous forme de listes de listes, de taille prédéfinie, remplie aléatoirement.
* Ecrire un produit matrice/vecteur utilisant cette structure et basée sur une double boucle.
* Comparer le temps d'exécution (package time) de votre produit avec la même opération réalisée par Numpy.

In [13]:
# Import de package 
import random  # https://docs.python.org/3.7/library/random.html
import time    # https://docs.python.org/3.7/library/time.html

In [45]:
def afficheTemps(func, args=()):
    tmp1 = time.process_time()
    val = func(*args)
    tmp = time.process_time() - tmp1
    print("Temps exécution (en secondes):", tmp)
    return val

In [33]:
def soluceCreationMatrice(ligne,colonne):
    return [[random.randint(0,10) for i in range(colonne)] for j in range(ligne)]

In [25]:
def createMatrice(sizeN, sizeM, entreA, etB):
    ret = []
    for i in range(0,sizeN):
        ligne = []
        for j in range(0,sizeM):
            ligne.append(random.randint(entreA,etB))
        ret.append(ligne)
    return ret

In [32]:
print(createMatrice(2,3,1,20))
print(soluceCreationMatrice(3,3))

[[3, 5, 9], [4, 20, 12]]
[[1, 1, 9], [6, 9, 3], [0, 0, 1]]


In [29]:
def produit(matrice, vecteur):
    if(len(matrice[0]) != len(vecteur)):
        return 0
    ret = []
    for i in range(0,len(matrice)):
        somme = 0
        for j in range(0,len(matrice[i])):
            somme = somme + matrice[i][j] * vecteur[j]
        ret.append(somme)
    return ret

In [56]:
mat = createMatrice(1000,1000,1,2000000000)
vecteurProd = [random.randint(0,3000000000) for i in range(1000)]
afficheTemps(produit,(mat,vecteurProd))
afficheTemps(np.dot,(mat,vecteurProd))
print("resultat")

Temps exécution (en secondes): 0.1086162629999996
Temps exécution (en secondes): 0.08378199599999991
resultat


Quel temps obtenez vous ? Les résultats son en secondes. On peut voir que numpy est optimisé pour les grandes données

Cas | $(m,n)$ | Boucle for | Solution Numpy
: | :-: | :-: | :-:
Cas 1 | (10,10) | 2.715400000008472e-05|7.525899999993868e-05
Cas 2 | (100,100)|0.001132144000000057|0.0014082020000003581
Cas 3 | (1000,1000)| 0.11101259399999996|0.08378199599999991

## Exercice 2 : pickle

* Définissez votre propre liste de chaînes de caractère (par exemple, une liste de villes de Bretagne).
* Utilisez le module ``pickle`` pour sauver cette liste dans un fichier texte : [Pickle](https://docs.python.org/3/library/pickle.html).
* Rechargez cette liste.
* Même chose avec un dictionnaire.

In [58]:
import pickle

In [60]:
list_ville = { "Vannes", "Rennes", "Brest","Saint-Brieuc", "Le Bono" }

In [61]:
pickle.dump( list_ville, open( "save.p", "wb" ) )

In [63]:
listRecupere = pickle.load( open( "save.p", "rb" ) )
print(listRecupere)

{'Brest', 'Saint-Brieuc', 'Le Bono', 'Rennes', 'Vannes'}


In [65]:
mon_dictionnaire = {}
mon_dictionnaire["nom"] = "Roger" 
mon_dictionnaire["espece"] = "kangourou"
mon_dictionnaire["poids"] = "89 kg"
mon_dictionnaire["taille"] = "2,07 m"
mon_dictionnaire["corpulence"] = "bodybuilder"

pickle.dump( mon_dictionnaire, open( "animaux_connus.p", "wb" ) )

listAnimaux = pickle.load(open("animaux_connus.p", "rb"))

print(listAnimaux)

{'nom': 'Roger', 'espece': 'kangourou', 'poids': '89 kg', 'taille': '2,07 m', 'corpulence': 'bodybuilder'}


## Exercice 3 : Classe

* Créer une classe Python nommée ``Vecteur``, que l'on utilisera pour décrire un vecteur plan. Ces paramètres sont ``x`` et ``y``.
* Implémenter l'addition ``__add__``, la soustraction ``__sub__``, les multiplications ``__mul__`` et ``__rmul__`` par un vecteur et un scalaire, et la représentation ``__repr__``. Ne pas oublier le constructeur ``__init__``.
* Créer une classe nommée ``Particule``, dont les paramètres sont les vecteurs position et vitesse. Implémenter une méthode ``deplace(dt)`` pour calculer la nouvelle position d'une particule après un pas de temps dt.

In [24]:
class Vecteur:    
    def __init__(self, x_, y_):
        self.x = x_
        self.y = y_
    
    def __add__(self, vecteur):
        return Vecteur(self.x + vecteur.x, self.y + vecteur.y)
    
    def __sub__(self, vecteur):
        return Vecteur(self.x - vecteur.x, self.y - vecteur.y)
    
    def __mul__(self, vecteur):
        return self.x * vecteur.x + self.y * vecteur.y
    
    def __rmul__(self, scalaire):
        return Vecteur(self.x * scalaire, self.y * scalaire)
    
    def __rep__(self):
        return {'x':self.x, 'y':self.y}
    
    def __str__(self):
        return "[%s,%s]" % (self.x, self.y)
    
class Particule:    
    def __init__(self, position_, vitesse_):
        self.position = position_
        self.vitesse = vitesse_
    
    def deplace(self,dt):
        self.position = self.position + (dt * self.vitesse )
    
    def __str__(self):
        return "Particule(%s,%s)" % (self.position, self.vitesse)

In [18]:
vect1 = Vecteur(2,4)
vect2 = Vecteur(1,1)
vect3 = vect1 + vect2
print(vect3)
vect4 = vect1 - vect2
print(vect4)
print(vect1 * vect2)
print( 3 * vect1)

[3,5]
[1,3]
6
[6,12]


Testez le bon fonctionnement de vos deux classes

In [27]:
vecPos = Vecteur(1,1)
vecVitesse = Vecteur(3,2)
dt = 5
particule = Particule(vecPos, vecVitesse)
print(particule)
particule.deplace(1)
print(particule)
particule.deplace(-0.5)
print(particule)

Particule([1,1],[3,2])
Particule([4,3],[3,2])
Particule([2.5,2.0],[3,2])


## Exercice 4 : Simpy

$$ I = \int_{-\infty}^{+\infty} e^{-x^2}dx$$

* Calculez l’intégrale suivante (dite de Gauss) en utilisant la méthode symbolique de [Simpy](https://docs.sympy.org/1.5.1/modules/integrals/integrals.html) 
* Calculez cette même intégrale numériquement en utilisant la méthode des trapèzes de numpy (i.e. ``numpy.trapz``) 
* Calculez une nouvelle fois la même intégrale avec une méthode de calcul équivalente que vous aurez codée.



In [45]:
#!pip install simpy
from simpy import *
from sympy import *
import math
x = symbols('x')

In [48]:
#inte = integrate(math.exp(- math.pow(x,2), (x, - np.inf, np.inf)))
integrate(x**2 * exp(x) * cos(x), x)

x**2*exp(x)*sin(x)/2 + x**2*exp(x)*cos(x)/2 - x*exp(x)*sin(x) + exp(x)*sin(x)/2 - exp(x)*cos(x)/2

## Exercice 5 : image

* Récupérez l’image hyperspectrale nommée ``Urban_R162.mat`` 
* Examinez la structure de données correspondantes 
* Faire une fonction qui affiche une bande spectrale donnée de l’image sous la forme d’une image avec une colormap 
* Affichez une composition couleur de cette image en choisissant la correspondance spectre/couleur suivante : 

 50->R 100->G 150->B 
* Affichez sur un même diagramme 10 spectres pris aléatoirement dans l’image

In [41]:
%matplotlib inline
import matplotlib.pyplot as plt

In [2]:
# Lecture d'un fichier .mat en Python
import scipy.io
mat = scipy.io.loadmat('Urban_R162.mat')

Comment est organisé la donnée (quelle structure ?) ? Quelle est la taille de l'image ? [A compléter]